src/lib.rs

```diff
@@ -135,6 +136,68 @@ pub struct GptDisk {
 }
 
 impl GptDisk {
+    /// Add another partition to this disk.  This tries to find
+    /// the optimum partition location with the lowest block device
+    pub fn add_partition(
+        &mut self,
+        name: &str,
+        size: usize,
+        part_type: partition::PartitionType,
+        flags: u64,
+    ) -> io::Result<()> {
+        self.sort_partitions();
+        // Find the lowest lba that is larger than size
+        let free_sections = self.find_free_sectors();
+        for (starting_lba, length) in free_sections {
+            if length as usize >= size {
+                // Found our free slice
+                let part = partition::Partition {
+                    part_type_guid: part_type,
+                    part_guid: uuid::Uuid::new_v4(),
+                    first_lba: starting_lba,
+                    last_lba: starting_lba + size as u64,
+                    flags,
+                    name: name.to_string(),
+                };
+                self.partitions.push(part);
+                return Ok(());
+            }
+        }
+
+        Err(io::Error::new(
+            io::ErrorKind::Other,
+            "Unable to find enough space on drive",
+        ))
+    }
+
+    /// Find free space on the disk
+    /// Returns a tuple of (starting_lba, length in lba's)
+    pub fn find_free_sectors(&self) -> Vec<(u64, u64)> {
+        if let Some(primary_header) = self.primary_header() {
+            println!("first_usable: {}", primary_header.first_usable);
+            let mut disk_positions = vec![primary_header.first_usable + 1];
+            for part in self.partitions().iter().filter(|p| p.is_used()) {
+                println!("partition: ({}, {})", part.first_lba, part.last_lba);
+                disk_positions.push(part.first_lba);
+                disk_positions.push(part.last_lba);
+            }
+            disk_positions.push(primary_header.last_usable - 1);
+            println!("last_usable: {}", primary_header.last_usable);
+            disk_positions.sort();
+
+            return disk_positions
+                // Walk through the LBA's in chunks of 2 (ending, starting)
+                .chunks(2)
+                // Add 1 to the ending and then subtract the starting
+                .map(|p| (p[0] + 1, p[1] - 1))
+                // Ensure that none of these blocks is negative
+                .filter(|(_, space)| *space > 0)
```

Comment:
If I'm reading this correctly, this is already a `u64`, so it can't be negative.

Reply:
Good point

Reply:
Good point.  I'l remove this filter now that i have saturating_sub